File: alura/php-design-patterns/EmAprovacao.php
<?php
  class EmAprovacao implements EstadoDeUmOrcamento {
    private $descontoAplicado = false;

    public function aplicaDescontoExtra(Orcamento $orcamento) {
      if(!$this->descontoAplicado) {
        $orcamento->setValor($orcamento->getValor() - $orcamento->getValor() * 0.02);
        $this->descontoAplicado = true;
      } else {
        throw new Exception("Desconto já aplicado!");
      }
    }

    public function aprova(Orcamento $orcamento) {
      $orcamento->setEstado(new Aprovado());
    }

    public function reprova(Orcamento $orcamento) {
      $orcamento->setEstado(new Reprovado());
    }

    public function finaliza(Orcamento $orcamento)  {
      throw new Exception ("Apenas orçamentos aprovados podem ser finalizados");
    }
  }
?>

File: alura/c-sharp/design-patterns-1/EmAprovacao.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CursoDesignPatterns
{
  public class EmAprovacao : EstadoDeUmOrcamento
  {
    public void AplicaDescontoExtra(Orcamento orcamento)
    {
      orcamento.Valor = orcamento.Valor - (orcamento.Valor * 0.05);
    }

    public void Aprova(Orcamento orcamento) {
      orcamento.EstadoAtual = new Aprovado();
    }

    public void Reprova(Orcamento orcamento) {
      orcamento.EstadoAtual = new Reprovado();
    }

    public void Finaliza(Orcamento orcamento) {
      throw new Exception("Orcamento em aprovacao não pode ir para finalizado direto");
    }
  }
}
